File: app/src/main/java/com/ladsoft/bakingapp/fragment/RecipeFragment.java
package com.ladsoft.bakingapp.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.ladsoft.bakingapp.R;
import com.ladsoft.bakingapp.adapter.RecipeIngredientsAdapter;
import com.ladsoft.bakingapp.adapter.RecipeStepsAdapter;
import com.ladsoft.bakingapp.entity.Recipe;

import butterknife.BindView;
import butterknife.ButterKnife;

public class RecipeFragment extends Fragment {
    @BindView(R.id.ingredients) RecyclerView ingredients;
    @BindView(R.id.steps) RecyclerView steps;

    private RecipeIngredientsAdapter ingredientsAdapter;
    private RecipeStepsAdapter stepsAdapter;
    private RecipeStepsAdapter.Listener listener;

    public static RecipeFragment newInstance() {
        return new RecipeFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_recipe, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        Context context = getContext();
        ingredients.setNestedScrollingEnabled(false);
        ingredients.setLayoutManager(new LinearLayoutManager(context));
        ingredientsAdapter = new RecipeIngredientsAdapter(LayoutInflater.from(context));
        ingredients.setAdapter(ingredientsAdapter);

        steps.setNestedScrollingEnabled(false);
        steps.setLayoutManager(new LinearLayoutManager(context));
        steps.addItemDecoration(new DividerItemDecoration(context, DividerItemDecoration.VERTICAL));
        stepsAdapter = new RecipeStepsAdapter(LayoutInflater.from(context));
        stepsAdapter.setListener(listener);
        steps.setAdapter(stepsAdapter);
    }

    public void setListener(RecipeStepsAdapter.Listener listener) {
        this.listener = listener;
    }

    public void setDatasource(Recipe recipe) {
        ingredientsAdapter.setDatasource(recipe.getIngredients());
        stepsAdapter.setDatasource(recipe.getSteps());
    }
}

File: app/src/main/java/com/ladsoft/bakingapp/fragment/RecipeFragment.kt
package com.ladsoft.bakingapp.fragment


import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.ladsoft.bakingapp.R
import com.ladsoft.bakingapp.adapter.RecipeIngredientsAdapter
import com.ladsoft.bakingapp.adapter.RecipeStepsAdapter
import com.ladsoft.bakingapp.entity.Recipe

import butterknife.BindView
import butterknife.ButterKnife

class RecipeFragment : Fragment() {

    @JvmField @BindView(R.id.ingredients)
    internal var ingredients: RecyclerView? = null

    @JvmField @BindView(R.id.steps)
    internal var steps: RecyclerView? = null

    private lateinit var ingredientsAdapter: RecipeIngredientsAdapter
    private lateinit var stepsAdapter: RecipeStepsAdapter
    private lateinit var listener: RecipeStepsAdapter.Listener

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_recipe, container, false)
        ButterKnife.bind(this, view)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        ingredients!!.isNestedScrollingEnabled = false
        ingredients!!.layoutManager = LinearLayoutManager(context)
        ingredientsAdapter = RecipeIngredientsAdapter(LayoutInflater.from(context))
        ingredients!!.adapter = ingredientsAdapter

        steps!!.isNestedScrollingEnabled = false
        steps!!.layoutManager = LinearLayoutManager(context)
        steps!!.addItemDecoration(DividerItemDecoration(context!!, DividerItemDecoration.VERTICAL))
        stepsAdapter = RecipeStepsAdapter(LayoutInflater.from(context))
        stepsAdapter.setListener(listener)
        steps!!.adapter = stepsAdapter
    }

    fun setListener(listener: RecipeStepsAdapter.Listener) {
        this.listener = listener
    }

    fun setDatasource(recipe: Recipe?) {
        ingredientsAdapter.setDatasource(recipe?.ingredients?.toMutableList())
        stepsAdapter.setDatasource(recipe?.steps?.toMutableList())
    }

    companion object {

        fun newInstance(): RecipeFragment {
            return RecipeFragment()
        }
    }
}
